common/rfb/CSecurityRSAAES.cxx

```diff
@@ -0,0 +1,444 @@
+    throw AuthFailureException("gnutls_global_init failed");
+  try {
+    writePublicKey();
+  } catch (...) {
+    cleanup();
+    throw;
+  }
+  if (secType == secTypeRA2 || secType == secTypeRA2ne)
+    keySize = 16;
+  else
+    keySize = 32;
+}
+
+CSecurityRSAAES::~CSecurityRSAAES()
+{
+  cleanup();
+}
+
+void CSecurityRSAAES::cleanup()
+{
+  gnutls_global_deinit();
+  if (serverKeyN)
+    delete[] serverKeyN;
+  if (serverKeyE)
+    delete[] serverKeyE;
+  if (clientKeyN)
+    delete[] clientKeyN;
+  if (clientKeyE)
+    delete[] clientKeyE;
+  if (clientKey.size)
+    rsa_private_key_clear(&clientKey);
+  if (clientPublicKey.size)
+    rsa_public_key_clear(&clientPublicKey);
+  if (serverKey.size)
+    rsa_public_key_clear(&serverKey);
+  if (secType == secTypeRA2 || secType == secTypeRA256)
+    cc->setStreams(rawis, rawos);
+  if (rais)
+    delete rais;
+  if (raos)
+    delete raos;
+}
+
+bool CSecurityRSAAES::processMsg()
+{
+  switch (state) {
+    case ReadPublicKey:
+      if (readPublicKey()) {
+        verifyServer();
+        writeRandom();
+        state = ReadRandom;
+      }
+      return false;
+    case ReadRandom:
+      if (readRandom()) {
+        setCipher();
+        writeHash();
+        state = ReadHash;
+      }
+      return false;
+    case ReadHash:
+      if (readHash())
+        state = ReadSubtype;
+    case ReadSubtype:
+      if (readSubtype()) {
+        writeCredentials();
+        return true;
+      }
+      return false;
+  }
+  throw AuthFailureException("unreachable");
+}
+
+const char* CSecurityRSAAES::description() const
+{
+  switch (secType) {
+    case secTypeRA2:      return "RSA-AES security type";
+    case secTypeRA2ne:    return "RSA-AES unencrypted security type";
+    case secTypeRA256:    return "RSA-AES-256 security type";
+    case secTypeRAne256:  return "RSA-AES-256 unencrypted security type";
+  }
+  return NULL;
+}
+
+static void random_func(void* ctx, size_t length, uint8_t* dst)
+{
+  gnutls_rnd(GNUTLS_RND_RANDOM, dst, length);
+}
+
+void CSecurityRSAAES::writePublicKey()
+{
+  // generate client key
+  rsa_public_key_init(&clientPublicKey);
+  rsa_private_key_init(&clientKey);
+  int rsaKeySize = (clientKeyLength + 7) / 8;
+  // set key size to non-zero to allow clearing the keys when cleanup
+  clientPublicKey.size = rsaKeySize;
+  clientKey.size = rsaKeySize;
+  // set e = 65537
+  mpz_set_ui(clientPublicKey.e, 65537);
+  if (!rsa_generate_keypair(&clientPublicKey, &clientKey,
+    NULL, random_func, NULL, NULL, clientKeyLength, 0))
+    throw AuthFailureException("failed to generate key");
+  clientKeyN = new rdr::U8[rsaKeySize];
+  clientKeyE = new rdr::U8[rsaKeySize];
+  nettle_mpz_get_str_256(rsaKeySize, clientKeyN, clientPublicKey.n);
+  nettle_mpz_get_str_256(rsaKeySize, clientKeyE, clientPublicKey.e);
+  rawos->writeU32(clientKeyLength);
+  rawos->writeBytes(clientKeyN, rsaKeySize);
+  rawos->writeBytes(clientKeyE, rsaKeySize);
+  rawos->flush();
```

Comment:
I would prefer you use normal stream names until the switch. This gives the impression that encryption is already active, and these writes are actively bypassing it.